@harshanarayana I repeated the test 300 times again and got 6 failures. Indeed, it might be the timeout issue which manifests on my machine. I include some logs from one failure in case they tell you something. 
```
    job_controller_test.go:1863: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-06-06 20:33:48.472283094 +0200 CEST m=+243.832841058 <nil> 0 0 0  nil 0xc0009dc354}
I0606 20:33:49.472060 3892298 job_controller.go:204] Shutting down job controller
I0606 20:33:49.472231 3892298 job_controller.go:190] Starting job controller
I0606 20:33:49.472254 3892298 shared_informer.go:255] Waiting for caches to sync for job
I0606 20:33:49.472266 3892298 shared_informer.go:262] Caches are synced for job
I0606 20:33:49.472361 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:49.472481 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:49.472524 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:50.472791 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:50.473648 3892298 job_controller.go:204] Shutting down job controller
E0606 20:33:50.473720 3892298 event.go:267] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"foobar.16f61bc160d90337", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"Job", Namespace:"default", Name:"foobar", UID:"d46f4daf-5314-42c0-b042-9fd6e01eb1b1", APIVersion:"", ResourceVersion:"", FieldPath:""}, Reason:"DeadlineExceeded", Message:"Job was active longer than specified deadline", Source:v1.EventSource{Component:"job-controller", Host:""}, FirstTimestamp:time.Date(2022, time.June, 6, 20, 33, 50, 472708919, time.Local), LastTimestamp:time.Date(2022, time.June, 6, 20, 33, 50, 472708919, time.Local), Count:1, Type:"Warning", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'request namespace does not match object namespace, request: "" object: "default"' (will not retry!)
I0606 20:33:50.473339 3892298 event.go:294] "Event occurred" object="default/foobar" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
I0606 20:33:50.474101 3892298 job_controller.go:190] Starting job controller
I0606 20:33:50.474183 3892298 shared_informer.go:255] Waiting for caches to sync for job
I0606 20:33:50.474197 3892298 shared_informer.go:262] Caches are synced for job
I0606 20:33:50.474478 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:50.475207 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:50.475309 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:51.475874 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:51.475926 3892298 event.go:294] "Event occurred" object="default/foobar" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
E0606 20:33:51.475947 3892298 event.go:267] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"foobar.16f61bc19ca2977c", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"Job", Namespace:"default", Name:"foobar", UID:"2ce0d273-29d4-4be8-9d90-a7c22722f970", APIVersion:"", ResourceVersion:"", FieldPath:""}, Reason:"DeadlineExceeded", Message:"Job was active longer than specified deadline", Source:v1.EventSource{Component:"job-controller", Host:""}, FirstTimestamp:time.Date(2022, time.June, 6, 20, 33, 51, 475775356, time.Local), LastTimestamp:time.Date(2022, time.June, 6, 20, 33, 51, 475775356, time.Local), Count:1, Type:"Warning", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'request namespace does not match object namespace, request: "" object: "default"' (will not retry!)
I0606 20:33:51.674143 3892298 job_controller.go:204] Shutting down job controller
I0606 20:33:51.674318 3892298 job_controller.go:190] Starting job controller
I0606 20:33:51.674339 3892298 shared_informer.go:255] Waiting for caches to sync for job
I0606 20:33:51.674348 3892298 shared_informer.go:262] Caches are synced for job
I0606 20:33:51.674490 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:51.674596 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:51.674646 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:52.674872 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:52.674936 3892298 event.go:294] "Event occurred" object="default/foobar" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
E0606 20:33:52.674985 3892298 event.go:267] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"foobar.16f61bc1e4199b0e", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"Job", Namespace:"default", Name:"foobar", UID:"7d7bb219-9e93-4203-a107-e2d53a1690cf", APIVersion:"", ResourceVersion:"", FieldPath:""}, Reason:"DeadlineExceeded", Message:"Job was active longer than specified deadline", Source:v1.EventSource{Component:"job-controller", Host:""}, FirstTimestamp:time.Date(2022, time.June, 6, 20, 33, 52, 674757390, time.Local), LastTimestamp:time.Date(2022, time.June, 6, 20, 33, 52, 674757390, time.Local), Count:1, Type:"Warning", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'request namespace does not match object namespace, request: "" object: "default"' (will not retry!)
I0606 20:33:52.875345 3892298 job_controller.go:204] Shutting down job controller
I0606 20:33:52.875559 3892298 job_controller.go:190] Starting job controller
I0606 20:33:52.875596 3892298 shared_informer.go:255] Waiting for caches to sync for job
I0606 20:33:52.875607 3892298 shared_informer.go:262] Caches are synced for job
I0606 20:33:52.875833 3892298 job_controller.go:504] enqueueing job default/foobar
I0606 20:33:53.876095 3892298 job_controller.go:504] enqueueing job default/foobar
--- FAIL: TestSyncPastDeadlineJobFinished (2.00s)
```

--- a/pkg/controller/job/job_controller.go
+++ b/pkg/controller/job/job_controller.go
@@ -746,17 +746,11 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 	if feature.DefaultFeatureGate.Enabled(features.JobReadyPods) {
 		ready = pointer.Int32(countReadyPods(activePods))
 	}
-	// Job first start. Set StartTime and start the ActiveDeadlineSeconds timer
-	// only if the job is not in the suspended state.
+
+	// Job first start. Set StartTime only if the job is not in the suspended state.
 	if job.Status.StartTime == nil && !jobSuspended(&job) {
 		now := metav1.Now()
 		job.Status.StartTime = &now
-		// enqueue a sync to check if job past ActiveDeadlineSeconds
-		if job.Spec.ActiveDeadlineSeconds != nil {
-			klog.V(4).Infof("Job %s has ActiveDeadlineSeconds will sync after %d seconds",
-				key, *job.Spec.ActiveDeadlineSeconds)
-			jm.queue.AddAfter(key, time.Duration(*job.Spec.ActiveDeadlineSeconds)*time.Second)
-		}
 	}
 
 	var manageJobErr error
@@ -775,6 +769,10 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 		finishedCondition = newCondition(batch.JobFailed, v1.ConditionTrue, "BackoffLimitExceeded", "Job has reached the specified backoff limit")
 	} else if pastActiveDeadline(&job) {
 		finishedCondition = newCondition(batch.JobFailed, v1.ConditionTrue, "DeadlineExceeded", "Job was active longer than specified deadline")
+	} else if job.Spec.ActiveDeadlineSeconds != nil && !jobSuspended(&job) {
+		syncDuration := time.Duration(*job.Spec.ActiveDeadlineSeconds)*time.Second - time.Since(job.Status.StartTime.Time)
+		klog.V(2).InfoS("Job has activeDeadlineSeconds configuration. Will sync this job again", "job", key, "nextSyncIn", syncDuration)
+		jm.queue.AddAfter(key, syncDuration)
 	}
 
 	var prevSucceededIndexes, succeededIndexes orderedIntervals

--- a/pkg/controller/job/job_controller_test.go
+++ b/pkg/controller/job/job_controller_test.go
@@ -59,11 +59,11 @@ import (
 
 var alwaysReady = func() bool { return true }
 
-func newJob(parallelism, completions, backoffLimit int32, completionMode batch.CompletionMode) *batch.Job {
+func newJobWithName(name string, parallelism, completions, backoffLimit int32, completionMode batch.CompletionMode) *batch.Job {
 	j := &batch.Job{
 		TypeMeta: metav1.TypeMeta{Kind: "Job"},
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      "foobar",
+			Name:      name,
 			UID:       uuid.NewUUID(),
 			Namespace: metav1.NamespaceDefault,
 		},
@@ -103,6 +103,10 @@ func newJob(parallelism, completions, backoffLimit int32, completionMode batch.C
 	j.Spec.BackoffLimit = &backoffLimit
 
 	return j
+}
+
+func newJob(parallelism, completions, backoffLimit int32, completionMode batch.CompletionMode) *batch.Job {
+	return newJobWithName("foobar", parallelism, completions, backoffLimit, completionMode)
 }
 
 func newControllerFromClient(kubeClient clientset.Interface, resyncPeriod controller.ResyncPeriodFunc) (*Controller, informers.SharedInformerFactory) {
@@ -1828,40 +1832,66 @@ func hasTrueCondition(job *batch.Job) *batch.JobConditionType {
 }
 
 func TestSyncPastDeadlineJobFinished(t *testing.T) {
-	clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: "", ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+	clientset := fake.NewSimpleClientset()
 	manager, sharedInformerFactory := newControllerFromClient(clientset, controller.NoResyncPeriodFunc)
-	fakePodControl := controller.FakePodControl{}
-	manager.podControl = &fakePodControl
 	manager.podStoreSynced = alwaysReady
 	manager.jobStoreSynced = alwaysReady
-	var actual *batch.Job
-	manager.updateStatusHandler = func(ctx context.Context, job *batch.Job) (*batch.Job, error) {
-		actual = job
-		return job, nil
-	}
 
-	job := newJob(1, 1, 6, batch.NonIndexedCompletion)
-	activeDeadlineSeconds := int64(10)
-	job.Spec.ActiveDeadlineSeconds = &activeDeadlineSeconds
-	start := metav1.Unix(metav1.Now().Time.Unix()-15, 0)
-	job.Status.StartTime = &start
-	job.Status.Conditions = append(job.Status.Conditions, *newCondition(batch.JobFailed, v1.ConditionTrue, "DeadlineExceeded", "Job was active longer than specified deadline"))
-	sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(job)
-	forget, err := manager.syncJob(context.TODO(), testutil.GetKey(job, t))
-	if err != nil {
-		t.Errorf("Unexpected error when syncing jobs %v", err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	sharedInformerFactory.Start(ctx.Done())
+
+	go manager.Run(ctx, 1)
+
+	tests := []struct {
+		name         string
+		setStartTime bool
+		jobName      string
+	}{
+		{
+			name:         "New job created without start time being set",
+			setStartTime: false,
+			jobName:      "job1",
+		},
+		{
+			name:         "New job created with start time being set",
+			setStartTime: true,
+			jobName:      "job2",
+		},
 	}
-	if !forget {
-		t.Errorf("Unexpected forget value. Expected %v, saw %v\n", true, forget)
-	}
-	if len(fakePodControl.Templates) != 0 {
-		t.Errorf("Unexpected number of creates.  Expected %d, saw %d\n", 0, len(fakePodControl.Templates))
-	}
-	if len(fakePodControl.DeletePodName) != 0 {
-		t.Errorf("Unexpected number of deletes.  Expected %d, saw %d\n", 0, len(fakePodControl.DeletePodName))
-	}
-	if actual != nil {
-		t.Error("Unexpected job modification")
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			job := newJobWithName(tc.jobName, 1, 1, 6, batch.NonIndexedCompletion)
+			activeDeadlineSeconds := int64(1)
+			job.Spec.ActiveDeadlineSeconds = &activeDeadlineSeconds
+			if tc.setStartTime {
+				start := metav1.Unix(metav1.Now().Time.Unix()-1, 0)
+				job.Status.StartTime = &start
+			}
+
+			_, err := clientset.BatchV1().Jobs(job.GetNamespace()).Create(ctx, job, metav1.CreateOptions{})
+			if err != nil {
+				t.Errorf("Could not create Job: %v", err)
+			}
+
+			if err := sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(job); err != nil {
+				t.Fatalf("Failed to insert job in index: %v", err)
+			}
+			var j *batch.Job
+			err = wait.Poll(200*time.Millisecond, 3*time.Second, func() (done bool, err error) {
+				j, err = clientset.BatchV1().Jobs(metav1.NamespaceDefault).Get(ctx, job.GetName(), metav1.GetOptions{})
+				if err != nil {
+					return false, nil
+				}
+				if len(j.Status.Conditions) == 1 && j.Status.Conditions[0].Reason == "DeadlineExceeded" {
+					return true, nil
+				}
+				return false, nil
+			})
+			if err != nil {
+				t.Errorf("Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in %v", j.Status)
+			}
+		})
 	}
 }
 

